File: leetcode/easy/35.Search Insert Position/js/searchInsertPosition.js
"use strict";
// Search Insert Position

/**
 * 
 * @param {number[]} nums 
 * @param {number} target
 * @returns {number} 
 */

var searchInsertPosition = function(nums, target) {
    let leftPointer = 0;
    let rightPointer = nums.length - 1;
    let middlePointer = 0;
    while (leftPointer <= rightPointer) {
        middlePointer = Math.floor((leftPointer + rightPointer) / 2);
        if (nums[middlePointer] === target) {
            return middlePointer;
        } else if (nums[middlePointer] > target) {
            rightPointer = middlePointer;
            rightPointer--;
        } else if (nums[middlePointer] < target) {
            leftPointer = middlePointer;
            leftPointer++;
        };
    };
    if (target < nums[middlePointer] || target === nums[middlePointer]) {
        return middlePointer;
    } else {
        return middlePointer + 1;
    };
};

//Testing
let nums0 = [1,3,5,6];
let target0 = 10;
let nums1 = [1,3];
let target1 = 3;
console.log(searchInsertPosition(nums1, target1));

// Complexity Analysis
// Time complexity : O(logN).
// Space complexity: O(1)

File: leetcode/easy/35.Search Insert Position/python/searchInsertPosition.py
"""Search Insert Position"""

def search_insert_position(nums, target):
    def helper(arr, tar):
        left_pointer = 0
        right_pointer = len(arr) - 1
        while left_pointer <= right_pointer:
            middle_pointer = (left_pointer + right_pointer) // 2
            if arr[middle_pointer] == tar:
                return middle_pointer
            elif arr[middle_pointer] > tar:
                right_pointer = middle_pointer
                right_pointer -= 1
            elif arr[middle_pointer] < tar:
                left_pointer = middle_pointer
                left_pointer += 1
        return middle_pointer
    index = helper(nums, target)
    if target < nums[index] or target == nums[index]:
        return index
    else:
        return index + 1

#Testing
nums0 = [1,3,5,6]
target0 = 7
nums1 = [1,3]
target1 = 0
print(search_insert_position(nums0, target0))

'''Complexity Analysis
Time complexity : O(logN).
Space complexity: O(1)'''